cpp/include/raft/matrix/detail/linewise_op.cuh

```diff
@@ -0,0 +1,364 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cuda_utils.cuh>
+#include <raft/pow2_utils.cuh>
+#include <raft/vectorized.cuh>
+
+namespace raft {
+namespace matrix {
+namespace detail {
+
+template <typename Type, typename IdxType, std::size_t VecBytes, int BlockSize>
+struct Linewise {
+  static constexpr IdxType VecElems = VecBytes / sizeof(Type);
+
+  typedef raft::TxN_t<Type, VecElems> Vec;
+  typedef raft::Pow2<VecBytes> AlignBytes;
+  typedef raft::Pow2<VecElems> AlignElems;
+  typedef raft::Pow2<raft::WarpSize> AlignWarp;
+
+  template <typename Lambda, typename... Vecs>
+  static __device__ __forceinline__ void vectorCols(typename Vec::io_t* out,
+                                                    const typename Vec::io_t* in,
+                                                    const typename Vec::io_t* in_end,
+                                                    const IdxType rowLen,
+                                                    IdxType rowDiv,
+                                                    IdxType rowMod,
```

Comment:
I'd prefer to keep them in terms of `div`/`mod` and verbosely document what I mean. The point is, I tried to get away from row/col formulation to avoid the confusion of row-major vs col-major matrix layouts.

Reply:
Hmm, maybe I should rename the function then... :)